### PR TITLE
Add tests for the Knative Deployer controller

### PR DESCRIPTION
**Pull request by scothis**
Also enhances the InjectFailure method to handle more fine grain
failures, including specifying the error.

### CODE REVIEW

File: pkg/controllers/testing/client.go [hunk 1]
```diff
@@ -258,11 +260,58 @@ func (w *statusWriterWrapper) Patch(ctx context.Context, obj runtime.Object, pat
 //      // Makes calls to create stream return an error.
 //      rifftesting.InduceFailure("create", "Stream"),
 //   },
-func InduceFailure(verb, resource string) ReactionFunc {
+func InduceFailure(verb, kind string, o ...InduceFailureOpts) ReactionFunc {
+	opts := InduceFailureOpts{}
+	if len(o) > 0 {
```
Comment (sbawaska) [inline pkg/controllers/testing/client.go]: throw error on more than one InduceFailureOpts?

File: pkg/controllers/testing/client.go [hunk 2]
```diff
@@ -258,11 +260,58 @@ func (w *statusWriterWrapper) Patch(ctx context.Context, obj runtime.Object, pat
...
-		if !action.Matches(verb, resource) {
+		if !action.Matches(verb, kind) {
+			return false, nil, nil
+		}
+		if opts.Namespace != "" && opts.Namespace != action.GetNamespace() {
+			return false, nil, nil
+		}
+		if opts.Name != "" {
+			switch a := action.(type) {
+			case GetAction:
```
Comment (sbawaska) [inline pkg/controllers/testing/client.go]:
```suggestion
			case GetAction, PatchAction, DeleteAction:
```
multiple conditions can be combined.

File: pkg/controllers/testing/client.go [hunk 3]
```diff
@@ -258,11 +260,58 @@ func (w *statusWriterWrapper) Patch(ctx context.Context, obj runtime.Object, pat
...
+				if opts.Name == a.GetName() {
+					return false, nil, nil
+				}
+			case CreateAction:
```
Comment (sbawaska) [inline pkg/controllers/testing/client.go]:
```suggestion
			case CreateAction, UpdateAction:
```

File: pkg/controllers/testing/client.go [hunk 4]
```diff
@@ -258,11 +260,58 @@ func (w *statusWriterWrapper) Patch(ctx context.Context, obj runtime.Object, pat
...
+		if !action.Matches(verb, kind) {
+			return false, nil, nil
+		}
+		if opts.Namespace != "" && opts.Namespace != action.GetNamespace() {
+			return false, nil, nil
+		}
+		if opts.Name != "" {
+			switch a := action.(type) {
+			case GetAction:
+				if opts.Name == a.GetName() {
```
Comment (sbawaska) [inline pkg/controllers/testing/client.go]:
```suggestion
				if opts.Name != a.GetName() {
```
I think all the name checks be `!=`

File: pkg/controllers/testing/client.go (hunk 2 above)
Comment (scothis) [reply pkg/controllers/testing/client.go]: doesn't work with typed switches, but I can introduce an interface that abstracts the common bits

**Pull request merged**